{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3719fee2-05e9-4fee-8f25-c7e66b267a4f",
   "metadata": {},
   "source": [
    "# Part 5 Mapping and variant calling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ab197a9-81f9-4cb8-ae2d-96d4f80334fb",
   "metadata": {},
   "source": [
    "## 5.1 Read mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c32ff27d-cc36-49da-8aaf-efe2f61bdc28",
   "metadata": {},
   "source": [
    "### Exercise 5.1.1 Comparing Capsella rubella and Capsella grandiflora\n",
    "\n",
    "1. Align the two sequences to find conserved and divergent regions.\n",
    "\n",
    "2. Whole genome shotgun sequencing was performed using the Monte Gargano (Italy) reference strain. Reads were collected with Sanger, Roche 454 XLR and Illumina GAIIx machines. Assembly was performed with Arachne v.20071016. Annotation was performed using tophat 1.3.0 and cufflinks 1.0.3.\n",
    "\n",
    "3. Busco analysis of the assembly shows Single copy genes were present 98.3%, suggesting a high quality assembly.\n",
    "\n",
    "4. On NIH Genbank, accessions from the article: Entrez BioProject ID PRJNA13878"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36a40d78-74be-48d0-b38a-a1caf6f33feb",
   "metadata": {},
   "source": [
    "### Exercise 5.1.2 Read mapping\n",
    "\n",
    "1. Read mapping is aligning sequence reads to a reference genome. Blast takes one sequence and compares it to another sequence or a database, while read mapping takes millions of short reads (or long reads), and aligns them to a reference genome. Blast algorithm would be too slow to align the millions of reads.\n",
    "2.\n",
    "- FreeBayes\n",
    "- Jitterbug\n",
    "- Fermi-Kit\n",
    "- CommonLaw\n",
    "- MetaSV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83a20c8c-bd0b-4476-b34d-57fc0adac32a",
   "metadata": {},
   "source": [
    "### Exercise 5.1.3 Capsella rubella reference genome\n",
    "\n",
    "1. bwa installed in /programs\n",
    "\n",
    "2. gff3,rna,cds,protein,genome,seq-report.\n",
    "- gff3 describes the locations and types of genomic features (genes, mRNAs, exons, CDS, UTRs, repeats, etc.) on a reference genome.\n",
    "- rna denotes a FASTA file containing the full transcript sequences (mRNA)\n",
    "- CDS (Coding DNA Sequences): a FASTA file containing only the coding regions of each transcript\n",
    "- Protein means a FASTA file with the amino acid sequences translated from the CDS sequences for each predicted protein coding gene.\n",
    "- Genome is the FASTA file of the reference genome assembly itself\n",
    "  \n",
    "3. Genome FASTA and gff3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "743e81e3-f148-4942-a63d-e949e4ee880e",
   "metadata": {},
   "source": [
    "### Exercise 5.1.4 Read mapping against the reference genome\n",
    "\n",
    "1. ```\n",
    "    wget https://github.com/samtools/samtools/releases/download/1.22.1/samtools-1.22.1.tar.bz2\n",
    "    tar xvfj samtools-1.22.1.tar.bz2\n",
    "    cd samtools-1.22.1/\n",
    "    make\n",
    "   ```\n",
    "2. ```~/p5/010-run-bwa.sh```\n",
    "3. Output files are sam files. Sam files stand for Sequence Alignment/Map files.\n",
    "   The first part has metadata lines starting with @. @HD refers to file format/version info. @SQ refers to reads sequence names and lengths. @PG can refer to programs used to generate alignments.\n",
    "   The next segment contains the alignment information. It is one line per read. This one starts with the line:\n",
    "   ```\n",
    "   N73019:1:1:1103:6775#0  99      NW_006238923.1  1058971 60      120M    =       1059205 354     GAAGGTATGACTCTACACGTTTCATACCAATCTTTAGACTTTCTTCTGTGGTCTGCATTCTCAATCCTGGTACCTCTTTCCAAGGAATGTTGAATTTTTCTTTTCATATGTTATTTCCTA CCCACABDDCDAE=EFDFDFCDE=EBC?EEEFFFFEDFE??=?CCBEE@5EBCB?BCE5CEEA:FAD5A?5?A?>C@>>C>>->>@BBBB:AB9BA:B?@;-'5/5>>=>@@:B@EB5@@ NM:i:1  MD:Z:102C17     MC:Z:120M MQ:i:60 AS:i:115        XS:i:0\n",
    "    ```\n",
    "N73019:1:1:1103:6775#0 is the name of the read\n",
    "99 is the bitwise flag\n",
    "NW_006238923.1 is the reference sequence name this read maps to\n",
    "1058971 is the leftmost position of the alignment\n",
    "60 mapping quality\n",
    "120M is the CIGAR string, here means 120 match\n",
    "= RNEXT, The next read alignment, whether it is to the same reference or not, in paired end the next one is the mate, and here it maps to the same reference\n",
    "1059205 position of the mate read on the reference\n",
    "354 template length (distance between outer ends of the pair reads)\n",
    "Then the sequence of the read\n",
    "CCCACABDDCDAE=EFDFDFCDE=EBC?... Base quality scores in ASCII\n",
    "rest are optional tags\n",
    "\n",
    "4. RNAME gives the reference genome or which chromosome the read mapped to, and POS gives the exact position of the mapping."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d66bd58e-93bc-4871-a77f-30aecd461bd8",
   "metadata": {},
   "source": [
    "### Exercise 5.1.5 Explore the alignments\n",
    "\n",
    "1. Samtools is used for Reading/writing/editing/indexing/viewing SAM/BAM/CRAM format.\n",
    "2. installed under ~/programs\n",
    "3. BAM files are SAM files in binary form.\n",
    "4. https://www.htslib.org/doc/1.13/samtools-view.html, p5/020-sam2bam.sh, p5/021-samtools-index.sh\n",
    "5. p5/023-samtools-count.sh\n",
    "6. ```samtools flagstat bam_files/Cr1504.sorted.bam```\n",
    "7. 024-samtools-filter.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52251661-dd43-4e59-a738-668c040580ae",
   "metadata": {},
   "source": [
    "### Exercise 5.1.6 Visual inspection\n",
    "1. Samtools tview is an alignment viewing tool.\n",
    "2. samtools tview your_sorted.bam your_reference.fasta\n",
    "   To navigate in tview:\n",
    "    - Arrow keys: Move left/right (one column at a time) or up/down (one row at a time).\n",
    "    - g: Go to a specific genomic position (you will be prompted to enter the position).\n",
    "    - q: Quit tview.\n",
    "    - N: Switch to the next viewing mode.\n",
    "    - ?: Display help with more commands.\n",
    "   fasta files, Gene Annotation Files:GTF (.gtf) and GFF (gff), Alignment Files (bam), variant files(vcf) and others"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72de2e21-224f-4424-b1fb-9abb9d6cf0bf",
   "metadata": {},
   "source": [
    "### Exercise 5.1.8 Comparing raw and pre-processed reads\n",
    "1. p5/030-mapping-corrected"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1b67404-7011-42e4-9388-8a37e04272b7",
   "metadata": {},
   "source": [
    "### Exercise 5.1.11 Mapping whole genome data\n",
    "1. ~/p5/040-mapping-bowtie.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c087b25-3dd8-436d-89fd-22582de896b4",
   "metadata": {},
   "source": [
    "### Exercise 5.1.12 Data from different technologies\n",
    "1. Canu, Falcon, minimap/miniasm were the assemblers used.\n",
    "2. minimap2 -a ref.fa query.fq > alignment.sam from https://github.com/lh3/minimap2\n",
    "   \n",
    "4. Illumina MiSeq: High accuracy, short reads\n",
    "PacBio: Long reads, higher error rates\n",
    "Oxford Nanopore: Long reads, variable error rates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877cef5a-4360-45b3-b2df-8f8e02bcd102",
   "metadata": {},
   "source": [
    "## 5.2 Genomic variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d28c52d-0cbb-4c00-abbc-c383348a4bd4",
   "metadata": {},
   "source": [
    "### Exercise 5.2.1 Variant calling\n",
    "1. Variant calling analyzes sequencing data to detect true genetic changes while distinguishing them from sequencing errors or artifacts.\n",
    "   Types:\n",
    "   Single nucleotide polymorphism/variant (SNP/SNV): change of a single base in the DNA sequence (1 base)\n",
    "   Insertion/deletion: addition of a group of bases, can cause frame shift (1-20 bases)\n",
    "   Structural variants: deletions, duplications, inversions, translocations, >20 bases\n",
    "   Copy number variations: gain or loss of large DNA segments, affecting gene dosage (hundreds to millions of bases)\n",
    "   Multi-nucleotide Variant (MNV): multiple adjacent base changes that are inherited together, (contiguous 2+ bases)\n",
    "   Complex/Other Variants: \tIncludes mobile element insertions, gene conversions, and more\n",
    "2. bcftools mpileup -f reference.fa alignments.bam | bcftools call -mv -Ob -o calls.bcf\n",
    "   p5/050-variant-calling-samtools.sh\n",
    "3. the output is in vcf format. The first header line denotes the file format, and the second header line is as follows:\n",
    "   #CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  /mnt/silo/hts2024/dnarabaci/p5/bam_files/Cg926.sorted.bam       /mnt/silo/hts2024/dnarabaci/p5/bam_files/Cr1504.sorted.bam\n",
    "4. IGV\n",
    "5. Script p5/051... to generate hom_snps, then the following command gives 2855: grep -v '^#' hom_snps.vcf | wc -l \n",
    "   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2d38cb4-b3fd-48d4-8136-e9ac9e35c22b",
   "metadata": {},
   "source": [
    "### Exercise 5.2.2 Comparing variant callers\n",
    "\n",
    "1. Clair3, DeepVariant, Octopus, GATK, FreeBayes, and Strelka2 (From [Barbitoff et al, 2022](https://rdcu.be/eNuwr))\n",
    "2. In the article, they compared variant calling tools, and DeepVariant consistently performed the best with high robustness."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
