{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "799fbb47-20f9-495c-a8cf-ec31ee1a98ce",
   "metadata": {},
   "source": [
    "# 4.1 De novo genome assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4458a9f-0d51-4bfb-b4c9-312306d3706a",
   "metadata": {},
   "source": [
    "## 4.1.1 Assembly approaches\n",
    "\n",
    "### 1. Briefly describe overlap-layout-consensus assembly.\n",
    "The first step is to find overlapping regions on the contigs, and based on the overlap generate a layout on a graph, that is deciding on an order and configuration of the contigs. Then the assembly is completed by resolving discrepancies by coming to a consensus.\n",
    "\n",
    "### 2. What is a de Bruijn graph?\n",
    "It is a graphical representation of a sequence, and works by generating k-mers from the reads, which are assigned as nodes, and if there is an overlap of k-1 characters, the nodes are connected with an edge.\n",
    "\n",
    "### 3. How can you use a de Bruijn graph for de novo genome assembly?\n",
    "The sequencing reads will be used to generate the k-mers as nodes. Then, by traversing the graph, the full sequence can be found. It should be noted that this method collapses the repetitive regions.\n",
    "\n",
    "### 4.\n",
    "de Bruijn graphs are much more computationally efficient, as they avoid the need for all-vs-all read comparisons required by OLC, which is particularly important for high throughput sequencing. Thanks to this, it also features lower memory usage and faster assembly times. Furthermore, because de Bruijn graphs collapse identical k-mers into one node, it effectively handles the challenge caused by repetitive regions. They can be particularly advantageous when it comes to handling short and error prone sequences.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c1a130f-2523-4894-aa8a-c7c5ca6239cc",
   "metadata": {},
   "source": [
    "## 4.1.2 Data preprocessing\n",
    "\n",
    "### 1. Estimate the quality of raw data.\n",
    "\n",
    "```{bash}\n",
    "./programs/FastQC/fastqc -o p4/1-2-Data-Preprocessing//fastqc/ data/Capsella-sequencing/DNA-seq/rawdata/*.fq.gz\n",
    "multiqc -n p4/1-2-Data-Preprocessing/fastqc/*\n",
    "\n",
    "```\n",
    "multiqc to visualize multiple files at once together\n",
    "In the raw sequencing data quality scores are worse towards the end of the reads, average quality per read is low, and there are problematic duplication statistics.\n",
    "\n",
    "### 2.Do quality trimming of the rawdata. You may wanna test multiple parameters.\n",
    "```{bash}\n",
    "outdir=~/p4/1-2-Data-Preprocessing/subsets-phred33-trimmomatic\n",
    "\n",
    "mkdir -p $outdir $outdir/unpaired\n",
    "\n",
    "for sample in Cru Cgr; do\n",
    "\n",
    "    java -jar \\\n",
    "         /mnt/silo/hts2024/dnarabaci/programs/Trimmomatic/Trimmomatic-0.39/trimmomatic-0.39.jar \\\n",
    "         PE \\\n",
    "         -threads 4 \\\n",
    "         /mnt/silo/hts2024/data/Capsella-sequencing/DNA-seq/subsets-phred33/$sample\\_1.fq.gz \\\n",
    "         /mnt/silo/hts2024/data/Capsella-sequencing/DNA-seq/subsets-phred33/$sample\\_2.fq.gz \\\n",
    "         $outdir/$sample\\_1.fq.gz \\\n",
    "         $outdir/unpaired/$sample\\_1_unpaired.fq.gz \\\n",
    "         $outdir/$sample\\_2.fq.gz \\\n",
    "         $outdir/unpaired/$sample\\_2_unpaired.fq.gz \\\n",
    "         LEADING:20 \\\n",
    "         TRAILING:20 \\\n",
    "         SLIDINGWINDOW:20:15 \\\n",
    "         MINLEN:36\n",
    "\n",
    "done\n",
    "\n",
    "```\n",
    "### 3. Do error correction of the rawdata. You may wanna test multiple programs and/or parameters.\n",
    "\n",
    "\n",
    "```{bash}\n",
    "#!/bin/bash\n",
    "PATH=/mnt/silo/hts2024/dnarabaci/programs/bfc/bfc-master:$PATH\n",
    "\n",
    "outdir=~/p4/1-2-Data-Preprocessing/subsets-phred33-bfc\n",
    "\n",
    "\n",
    "for f in /mnt/silo/hts2024/data/Capsella-sequencing/DNA-seq/subsets-phred33/C*.fq.gz; do\n",
    "    sample=$(echo $f | sed 's/.*\\///' | sed 's/.fq.gz//')\n",
    "    echo $sample $f\n",
    "    bfc \\\n",
    "        -s 1m \\\n",
    "        -t 4 \\\n",
    "        $f \\\n",
    "        | gzip > $outdir/$sample.fq.gz\n",
    "\n",
    "done\n",
    "\n",
    "### Quality control\n",
    "mkdir -p $outdir/fastqc\n",
    "\n",
    "/mnt/silo/hts2024/dnarabaci/programs/FastQC/fastqc \\\n",
    "    -o $outdir/fastqc \\\n",
    "    $outdir/*.fq.gz\n",
    "\n",
    "```\n",
    "### 4. Do adapter removal followed by error correction.\n",
    "\n",
    "## 4.1.3 De novo genome assembly using a short read assembler\n",
    "### 1.\n",
    "SOAPdenovo2\n",
    "\n",
    "### 2.\n",
    "\n",
    "```{bash}\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "PATH=/mnt/silo/hts2024/dnarabaci/programs/miniconda/bin:$PATH\n",
    "\n",
    "k=51\n",
    "\n",
    "### C. rubella rawdata\n",
    "mkdir -p ~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-raw/K$k\n",
    "\n",
    "SOAPdenovo-63mer all \\\n",
    "  -s ~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-raw.conf \\\n",
    "  -K $k \\\n",
    "  -R \\\n",
    "  -o ~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-raw/K$k \\\n",
    "  1>~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-raw/K$k-assembly.log \\\n",
    "  2>~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-raw/K$k-assembly.err\n",
    "\n",
    "\n",
    "### C. rubella bfc error corrected\n",
    "mkdir -p ~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-bfc/K$k\n",
    "\n",
    "SOAPdenovo-63mer all \\\n",
    "  -s ~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-bfc.conf \\\n",
    "  -K $k \\\n",
    "  -R \\\n",
    "  -o ~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-bfc/K$k \\\n",
    "  1>~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-bfc/K$k-assembly.log \\\n",
    "  2>~/p4/1-3-denovo-genome-assembly-using-short-assembler/soapdenovo/Cru-bfc/K$k-assembly.err\n",
    "\n",
    "```\n",
    "\n",
    "### 3.\n",
    "\n",
    "Capsella rubella bfc corrected:\n",
    " Scaffold number                  37\n",
    " In-scaffold contig number        504\n",
    " Total scaffold length            989536\n",
    " Average scaffold length          26744\n",
    " Filled gap number                163\n",
    " Longest scaffold                 179362\n",
    " Scaffold and singleton number    124\n",
    " Scaffold and singleton length    1016497\n",
    " Average length                   8197\n",
    " N50                              95324\n",
    " N90                              15401\n",
    " Weak points                      0\n",
    "\n",
    "Capsella rubella raw:  \n",
    " Scaffold number                  41\n",
    " In-scaffold contig number        593\n",
    " Total scaffold length            988778\n",
    " Average scaffold length          24116\n",
    " Filled gap number                152\n",
    " Longest scaffold                 156601\n",
    " Scaffold and singleton number    208\n",
    " Scaffold and singleton length    1020498\n",
    " Average length                   4906\n",
    " N50                              65930\n",
    " N90                              15327\n",
    " Weak points                      0\n",
    "\n",
    "\n",
    "\n",
    "### 4\n",
    "with the corrected version, the number of scaffolds are lower, average length is higher, and N50 is better.\n",
    "\n",
    "### 5\n",
    "Capsella grandiflora raw:\n",
    " Scaffold number                  1286\n",
    " In-scaffold contig number        15737\n",
    " Total scaffold length            743873\n",
    " Average scaffold length          578\n",
    " Filled gap number                883\n",
    " Longest scaffold                 8562\n",
    " Scaffold and singleton number    13177\n",
    " Scaffold and singleton length    1817294\n",
    " Average length                   137\n",
    " N50                              290\n",
    " N90                              52\n",
    " Weak points                      0\n",
    "\n",
    "Capsella grandiflora bfc:\n",
    " Scaffold number                  1204\n",
    " In-scaffold contig number        14908\n",
    " Total scaffold length            726989\n",
    " Average scaffold length          603\n",
    " Filled gap number                901\n",
    " Longest scaffold                 8586\n",
    " Scaffold and singleton number    12342\n",
    " Scaffold and singleton length    1735882\n",
    " Average length                   140\n",
    " N50                              301\n",
    " N90                              52\n",
    " Weak points                      0\n",
    "\n",
    "C.rubella had a much lower number of scaffolds as well as longer scaffolds, suggesting a more contiguous assembly. A very important metric is also N50, which the C. rubella assembly had a much higher N50, suggesting a higher quality assembly.\n",
    "\n",
    "## 4.1.4 Evaluating your genome assembly\n",
    "\n",
    "### 1.\n",
    "NG50 scaffold length,NG50 contig length,  Amount of scaffold sequence gene-sized scaffolds (≥25 Kbp), CEGMA, number of 458 core genes mapped, Fosmid coverage, Fosmid validity, VFR tag scaffold summary score, Optical map data, Optical map data, REAPR summary score\n",
    "\n",
    "### 2.\n",
    "NG50 scaffold length,NG50 contig length,  Amount of scaffold sequence gene-sized scaffolds (≥25 Kbp), CEGMA, number of 458 core genes mapped, REAPR summary score apply to de novo assemblies.\n",
    "### 3. \n",
    "N50 is calculated by summing all sequence lengths, starting with the longest, and observing the length that takes the sum length past 50% of the total assembly length. NG50 is similar, it is calculated by sorting scaffolds from longest to shortest and finding the length where cumulative length reaches 50% of the expected genome size.\n",
    "### 4.\n",
    "NG50 here is normalized to an expected genome length instead of total assembly length, and makes N50 values comparable between assemblies of the same species (or genome size), even if one assembly is missing parts of the genome.\n",
    "\n",
    "## 4.1.5 Other assembly programs\n",
    "\n",
    "SPAdes (de novo assembler for microbial, but has other versions, still not very suited to repetitive plant genome)\n",
    "MaSuRCA: Hybrid de novo assembler, Designed for large genomes, integrates Illumina + long-read data\n",
    "MEGAHIT: Ultra-fast de novo assembler (k-mer based), Low memory usage, scalable to large datasets.\n",
    "Canu: Long-read assembler (PacBio / Nanopore), Handles noisy long reads, built-in error correction, good for high-contiguity assemblies.\n",
    "Flye: Long-read assembler, Very fast with long reads, works well with low-coverage data, good repeat resolution.\n",
    "wtdbg2 (Redbean) : Long-read assembler, Extremely fast, lightweight, generates contiguous assemblies quickly.\n",
    "Hifiasm: HiFi read assembler (PacBio CCS reads), Very accurate, haplotype-resolved assemblies, handles high ploidy.\n",
    "\n",
    "Gorman, Z., Chen, J., de Leon, A.A.P. et al. Comparison of assembly platforms for the assembly of the nuclear genome of Trichoderma harzianum strain PAR3. BMC Genomics 24, 454 (2023). https://doi.org/10.1186/s12864-023-09544-6\n",
    "\n",
    "## 4.1.7 Whole genome de novo assembly of Capsella rubella and Capsella\n",
    "\n",
    "### 1. \n",
    "MaSuRCA handles large eukaryotic genomes better than SOAPdenovo2, produces more contiguous assemblies, integrates error correction.\n",
    "### 2. \n",
    "Use only raw data for MaSuRCA.\n",
    "### 3.\n",
    "https://bioinformaticsworkbook.org/dataAnalysis/GenomeAssembly/Assemblers/MaSuRCA.html#gsc.tab=0\n",
    "\n",
    "```{bash}\n",
    "git clone https://github.com/ISUGIFsingularity/masurca.git\n",
    "\n",
    "  MASURCA masurca sr_config.txt\n",
    "  MASURCA ./assemble.sh\n",
    "\n",
    "## sr_config.txt\n",
    "DATA\n",
    "   PE = pa 250 50 SRR3166543_1.fastq  SRR3166543_2.fastq\n",
    "   PE = pb 250 50 SRR3157034_1.fastq  SRR3157034_2.fastq\n",
    "   JUMP = ja 8000 1600 SRR3156163_1.fastq  SRR3156163_2.fastq\n",
    "   JUMP = jb 20000 4000 SRR3156596_1.fastq  SRR3156596_2.fastq\n",
    "   PACBIO = SRR3405330.fastq\n",
    "   END\n",
    "\n",
    "   PARAMETERS\n",
    "   GRAPH_KMER_SIZE = auto\n",
    "   USE_LINKING_MATES = 0\n",
    "   LIMIT_JUMP_COVERAGE = 300\n",
    "   CA_PARAMETERS =  cgwErrorRate=0.15\n",
    "   KMER_COUNT_THRESHOLD = 1\n",
    "   NUM_THREADS = 28\n",
    "   JF_SIZE = 200000000\n",
    "   SOAP_ASSEMBLY=0\n",
    "   END\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2e30d2a-5f53-4940-b877-7f12cdd0b1d9",
   "metadata": {},
   "source": [
    "## 4.1.8 Evolution of whole genome assembly\n",
    "### 1.\n",
    "Genome assembly has changed a lot over the past twenty years. The Human Genome Project (Frazier et al., 2001) used Sanger sequencing with a BAC-by-BAC approach, which was slow, labor-intensive, and left many gaps, especially in repetitive regions. By the time the panda genome was sequenced (Li et al., 2010), short-read next-generation sequencing made assemblies faster and cheaper, but they were still fragmented and struggled with repeats. More recently, genomes like those of mosquitoes (Matthews et al., 2018) and complex crops (Jarvis et al., 2017; Ling et al., 2018) have been assembled using long-read technologies, often combined with Hi-C or optical mapping, allowing for near-complete, chromosome-level assemblies that resolve difficult repetitive and structural regions. In short, genome assembly has moved from slow, gap-filled drafts to highly contiguous, high-quality genomes thanks to advances in sequencing and assembly strategies.\n",
    "\n",
    "### 2.\n",
    "By today, genome assembly has become far more accurate and complete than ever before. In this rapidly evolving field, higher computational power allowed with newer technology allows better algorithms to come to life that would not be possible before. High-fidelity long reads from PacBio HiFi and ultra-long Oxford Nanopore sequencing make it possible to read long stretches of DNA with very few errors. These reads are often combined with Hi-C or optical mapping data to build chromosome-level assemblies that are phased, meaning the two copies of each chromosome can be distinguished. Automated tools like hifiasm, Verkko, and Shasta make the process much faster and easier, even for complex genomes. Today’s assemblies can capture not just the genes but also repetitive regions, centromeres, telomeres, and structural variations, giving a near-complete picture of the genome.\n",
    "\n",
    "### 3.\n",
    "Bacterial genome assembly is now much easier than for complex eukaryotes. Long reads from PacBio or Oxford Nanopore often allow entire chromosomes and plasmids to be assembled in a single, gapless contig. Because bacterial genomes are small, circular, and haploid, they don’t require phasing, but repetitive elements, small plasmids, and mixed-strain samples can still pose challenges. Overall, complete bacterial genomes are now routine, though tricky cases need careful sequencing and analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20bbb241-73eb-488c-8dba-79018436f614",
   "metadata": {},
   "source": [
    "## 4.1.9 De novo genome assembly with long reads\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "090e7f0b-f0cf-49eb-96d3-24a553120847",
   "metadata": {},
   "source": [
    "## 4.2 Structural and functional gene annotation\n",
    "### 4.2.1 Methods of gene predictions\n",
    "\n",
    "#### 1. \n",
    "A gene is a stretch of DNA that carries instructions for making a protein or functional RNA. Functionally, it’s defined by the role its product plays in the cell, shaping traits and processes. From an evolutionary view, a gene is a heritable unit that can change over generations, influencing an organism’s survival and evolution.\n",
    "#### 2.\n",
    "Extrinsic and intrinsic methods. There are ab initio (intrinsic) methods without prior knowledge, and homology based (extrinsic) with information from homologs.  \n",
    "#### 3. \n",
    "Cufflinks, Scripture, SGP2 intrinsic, while GeneID, Augustus and Scripture are extrinsic.\n",
    "\n",
    "\n",
    "### 4.2.2 Aligning scaffolds against close species\n",
    "\n",
    "#### 1.\n",
    "Arabidopsis thaliana is close to these.\n",
    "#### 2.\n",
    "```\n",
    "#!/bin/bash\n",
    "\n",
    "PATH=/mnt/silo/hts2024/dnarabaci/programs/canu-2.2/canu/bin:$PATH\n",
    "\n",
    "canu \\\n",
    "    -p unknown \\\n",
    "    -d assembly \\\n",
    "    useGrid=false \\\n",
    "    genomeSize=1.5m \\\n",
    "    -pacbio /mnt/silo/hts2024/data/unknown/unknown-pacbio.fq.gz\n",
    "```\n",
    "\n",
    "### 4.2.3 Gene predictions using GeneMark\n",
    "\n",
    "#### 1. \n",
    "GeneMark-ES can be used for novel eukaryotic genomes.\n",
    "\n",
    "### 4.2.4 Sequencing based gene predictions\n",
    "Trinity and Oases, trinity modules can assemble contigs via kmers, or build de-Brujin graphs modeling gene isoforms or into actual transcripts, and Oases builds de_bruijn graphs from reads using different kmers for each graphs, then merge all assemblies into one.\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
